apiserver/facades/client/controller/controller.go

```diff
@@ -234,145 +233,15 @@ func (c *ControllerAPI) MongoVersion(ctx context.Context) (params.StringResult,
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	modelName := model.Name()
-	ctrCfg, err := c.controllerConfigService.ControllerConfig(ctx)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	controllerName := ctrCfg.ControllerName()
-
-	return &params.DashboardConnectionSSHTunnel{
-		Model:  fmt.Sprintf("%s:%s", controllerName, modelName),
-		Entity: fmt.Sprintf("%s/leader", appName),
-		Host:   fmt.Sprintf("%s", addr),
-		Port:   fmt.Sprintf("%d", port),
-	}, nil
-}
-
 // DashboardConnectionInfo returns the connection information for a client to
 // connect to the Juju Dashboard including any proxying information.
-func (c *ControllerAPI) DashboardConnectionInfo(ctx context.Context) (params.DashboardConnectionInfo, error) {
-	getDashboardInfo := func() (params.DashboardConnectionInfo, error) {
-		rval := params.DashboardConnectionInfo{}
-		controllerApp, err := c.state.Application(bootstrap.ControllerApplicationName)
-		if err != nil {
-			return rval, errors.Trace(err)
-		}
-
-		rels, err := controllerApp.Relations()
-		if err != nil {
-			return rval, errors.Trace(err)
-		}
-
-		for _, rel := range rels {
-			ep, err := rel.Endpoint(controllerApp.Name())
-			if err != nil {
-				return rval, errors.Trace(err)
-			}
-			if ep.Name != "dashboard" {
-				continue
-			}
-
-			model, ph, err := c.statePool.GetModel(rel.ModelUUID())
-			if err != nil {
-				return rval, errors.Trace(err)
-			}
-			defer ph.Release()
-
-			relatedEps, err := rel.RelatedEndpoints(controllerApp.Name())
-			if err != nil {
-				return rval, errors.Trace(err)
-			}
-			related := relatedEps[0]
-
-			appSettings, err := rel.ApplicationSettings(related.ApplicationName)
-			if err != nil {
-				return rval, errors.Trace(err)
-			}
-
-			if model.Type() != state.ModelTypeCAAS {
-				sshConnection, err := c.dashboardConnectionInfoForIAAS(
-					ctx,
-					related.ApplicationName,
-					appSettings)
-				rval.SSHConnection = sshConnection
-				return rval, err
-			}
-
-			proxyConnection, err := c.dashboardConnectionInfoForCAAS(ctx, related.ApplicationName)
-			rval.ProxyConnection = proxyConnection
-			return rval, err
-		}
-
-		return rval, errors.NotFoundf("dashboard")
-	}
-	conInfo, err := getDashboardInfo()
-
-	if conInfo.ProxyConnection != nil && conInfo.SSHConnection != nil {
-		return params.DashboardConnectionInfo{},
-			errors.New("cannot set both proxy and ssh connection for dashboard connection info")
-	}
-	conInfo.Error = apiservererrors.ServerError(err)
-	return conInfo, nil
+func (c *ControllerAPI) DashboardConnectionInfo(_ context.Context) (params.DashboardConnectionInfo, error) {
+	// TODO 27-01-2025 (hmlanigan)
+	// Reimplement the functionality in the controller charm.
+	// The most recent implementation used mongodb state, thus the method will
+	// now return Not Implemented rather than a temporary implementation with
+	// dqlite.
+	return params.DashboardConnectionInfo{}, errors.NotImplementedf("dashboard connection info")
```

Comment:
based on the output of `juju dashboard`, debating if this is the best error to return.

Reply:
I'm starting to think that all API calls should catch `NotImplemented`, which would allow us to deprecate endpoints more gracefully. It might be something that we feed into the new API work.

The code in question does trap the `NotFound`, but annoyingly doesn't allow you to pass the original server error message, which could supplement the error.

https://github.com/juju/juju/blob/317e2378d2107171c70fd9c3a9723358f60579c8/cmd/juju/dashboard/dashboard.go#L139-L147

Return `NotImplemented`, but put more information in the error message so it's not so cryptic.

Reply:
One difficulty with the NotImplemented is this built in attempt to help:

```
juju client not compatible with server: * (not implemented)
re-install your juju client to match the version running on the controller
```